File: python/host.py
import socket
from threading import Thread, Lock
import cv2
import base64
from json import dumps as dictToJson
from json import loads as jsonToDict
from json.decoder import JSONDecodeError

#################
### CONSTANTS ###
#################

ACK = 'ACK'
NEWLINE = '\n'.encode()
IMG_MSG_S = '{"type": "image", "data": "'.encode()
IMG_MSG_E = '"}'.encode()

ADDR = '127.0.0.1'
PORT = 8080
TIMEOUT = 2
SIZE = 256

########################
### CONNECTION CLASS ###
########################

class _connection:
	def __init__(self, socket, address, timeout, size):
		self.socket = socket
		self.address = address
		self.canWrite = True
		self.channels = {}
		self.timeout = timeout
		self.socket.settimeout(self.timeout)
		self.size = size
		self.stopped = False
		self.lock = Lock()
		self.start()

	def start(self):
		Thread(target=self.run, args=()).start()
		return self

	def run(self):
		tmp = ''
		while True:
			if self.stopped:
				self.socket.close()
				return

			try:
				tmp += self.socket.recv(self.size).decode()
				msg = jsonToDict(tmp)
				self.channels[msg['type']] = msg['data']
				if(msg['type'] == ACK):
					self.canWrite = True
				tmp = ''
			except JSONDecodeError: 
				continue
			except socket.timeout:
				continue

	def get(self, channel):
		if channel in self.channels.keys():
			return self.channels[channel]
		return None

	def encodeImg(self, img):
		success, encoded_img = cv2.imencode('.png', img)
		return base64.b64encode(encoded_img)

	def writeLock(self, channel, data):
		with self.lock:
			self.write(channel, data)

	def write(self, channel, data):
		if self.canWrite:
			self.canWrite = False
			msg = {'type': channel, 'data': data}
			self.socket.sendall(dictToJson(msg).encode() + NEWLINE)

	def writeImgLock(self, data):
		with self.lock:
			self.writeImg(data)

	def writeImg(self, data):
		if self.canWrite:
			self.canWrite = False
			msg = IMG_MSG_S + self.encodeImg(data) + IMG_MSG_E
			self.socket.sendall(msg + NEWLINE)

	def close(self):
		self.stopped = True

##################
### HOST CLASS ###
##################

class host:
	def __init__(self, addr=ADDR, port=PORT, timeout=TIMEOUT, size=SIZE, open=True):
		self.addr = addr
		self.port = port
		self.timeout = timeout
		self.size = size
		self.connections = []
		self.clients = []
		self.stopped = False
		self.opened = False
		if open:
			self.open()

	def set_timeout(self, time):
		self.timeout = time

	def open(self):
		while True:
			try:
				self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
				self.socket.settimeout(self.timeout)
				self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
				self.socket.bind(('', self.port))
				self.socket.listen()
				break
			except OSError as e:
				raise RuntimeError("Socket address in use: {}".format(e))
				return
			except socket.timeout:
				continue
		self.opened = True

	def start(self):
		Thread(target=self.run, args=()).start()
		return self

	def run(self):
		tmp = ''
		while True:
			if self.stopped:
				for c in self.clients:
					c.close()
				self.socket.close()
				return

			try:
				c, addr = self.socket.accept()
				if addr not in self.connections:
					self.connections.append(addr)
					self.clients.append(_connection(c, addr, self.timeout, self.size))
			except socket.timeout:
				continue


	def get_ALL(self, channel):
		data = []
		for c in self.clients:
			tmp = c.get(channel)
			if tmp is not None:
				data.append(tmp)
		return data

	def getClients(self):
		return self.clients

	def writeLock_ALL(self, channel, data):
		for c in self.clients:
			self.writeLock(channel, data)

	def write_ALL(self, channel, data):
		for c in self.clients:
			c.write(channel, data)

	def writeImgLock_ALL(self, data):
		for c in self.clients:
			c.writeImg(data)

	def writeImg_ALL(self, data):
		for c in self.clients:
			c.writeImg(data)

	def close(self):
		self.opened = False
		for c in self.clients:
			c.close()
		self.stopped = True

File: nodejs/host.js
'use strict'

var EventEmitter = require('events').EventEmitter;
var inherits = require('util').inherits;
var ip = require('ip');

/////////////////
/// CONSTANTS ///
/////////////////

var STOP = 'STOP';
var ACK = '__ack';
var IMAGE = 'image';
var NEWLINE = '\n';
var base64 = new RegExp('^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$');
var FAMILY = 'IPv4';

var ADDR = ip.address();
var PORT = 8080;
var MAXSIZE = 1500000;
var TIMEOUT = 0;

////////////////////////
/// CONNECTION CLASS ///
////////////////////////

function _connection(socket, address, timeout, maxSize) {
	EventEmitter.call(this);
	this.socket = socket;
	this.address = address;
	this.timeout = timeout;
	this.maxSize = maxSize;
	this.lastData = new Date().getTime();
	this.msgBuffer = '';
	this.channels = {};
	this.listener = null;

	//////////////////////
	/// SOCKET ACTIONS ///
	//////////////////////

	this.socket.on('data', (bytes) => {
		this.msgBuffer += bytes.toString();
		if(this.msgBuffer != '' && this.msgBuffer != '\n') {
			var data = this.msgBuffer.split('\n');
			for(var i = 0; i < data.length; i++) {
				try {
					if(data[i] == '') {
						continue;
					}
					var msg = JSON.parse(data[i]);
					this.channels[msg['type']] = msg['data'];
					if(msg['type'] == IMAGE) {
						if(base64.test(msg['data'])) {
							this.emit(msg['type'], msg['data']);
						}
					} else {
						this.emit(msg['type'], msg['data']);
					}
					this._reset();
					this.emit('data', msg);
				} catch(err) {};
			}
		}
	});
	this.socket.on('end', () => {
			this.emit('end');
		});

	this.socket.on('error', (err) => {
		this.emit('warning', err);
	});

	this._reset = function() {
		// this.msgBuffer = '';
		// this.lastData = new Date().getTime();
		// this.write(ACK, 'True');
		var op = null;
	}

	///////////////
	/// METHODS ///
	///////////////

	this.get = function(channel) {
		return this.channels[channel];
	}

	this.write = function(dataType, data) {
		var msg = {
			'type': dataType,
			'data': data
		};
		this.socket.write(JSON.stringify(msg) + NEWLINE);
	}

	this.close = function() {
		this.socket.destroy();
	}

	// Timeout handler
	if(this.timeout > 0) {
		setInterval( () => {
			if(((new Date().getTime() - this.lastData) / 1000) > this.timeout) {
				if(this.msgBuffer == '') {
					try { this.reset() }
					catch(err) {}
				}
			}
		}, this.timeout);
	}
}

inherits(_connection, EventEmitter);

//////////////////
/// HOST CLASS ///
//////////////////

function host(addr=ADDR, port=PORT, maxSize=MAXSIZE, timeout=TIMEOUT) {
	EventEmitter.call(this);
	this.net = require('net');

	this.addr = addr;
	this.port = port;
	this.timeout = timeout;
	this.maxSize = maxSize;

	this.clients = [];
	this.sockets = [];
	
	this.socketpath = {
		'port': this.port,
		'family': FAMILY,
		'address': this.addr
	};
	this.listener = null;
	this.opened = false;

	//////////////
	/// SERVER ///
	//////////////

	this.server = this.net.createServer((socket) => {
		this.listener = socket;
	});

	this.server.on('connection', (socket) => {
		for(var i = 0; i < this.sockets.length; i++) {
			if(this.sockets[i] === socket) {
				return
			}
		}
		this.sockets.push(socket);
		var connection = new _connection(socket, socket.address(), this.timeout, this.maxSize)
		this.clients.push(connection);
		connection.on('data', (msg) => {
			this.emit(msg['type'], this.get_ALL(msg['type']));
		});
		connection.on('error', (err) => {
			this.emit('warning', err);
		});
		connection.on('end', () => {
			this.emit('end');
		});
	});

	///////////////
	/// METHODS ///
	///////////////

	this.start = function() {
		this.server.listen(this.socketpath.port, this.socketpath.address);
		this.opened = true;
		return this;
	}

	this.get_ALL = function(channel) {
		var data = [];
		for(var i = 0; i < this.clients.length; i++) {
			var tmp = this.clients[i].get(channel)
			if(tmp != undefined) {
				data.push(tmp);
			}
		}
		return data;
	}

	this.getClients = function() {
		return this.clients;
	}

	this.write_ALL =function(channel, data) {
		for(var i = 0; i < this.clients.length; i++) {
			this.clients[i].write(channel, data);
		}
		return this;
	}

	this.close = function() {
		for(var i = 0; i < this.clients.length; i++) {
			this.clients[i].close();
		}
	}
}

inherits(host, EventEmitter);

module.exports = exports = host;
